spec/models/team_spec.rb

```diff
@@ -147,6 +147,86 @@
     end
   end
 
+  describe 'limit of coaches' do
+    let(:team) { create :team }
+    let(:new_user) { create :user }
+    let(:new_coach) { -> {{ name: 'coach', team_id: team.id, user_id: new_user.id }} }
+    let(:new_coach_as_student) { -> {{ name: 'student', team_id: team.id, user_id: new_user.id }} }
+    let(:second_new_coach) { -> {{ name: 'coach', team_id: team.id, user_id: create(:user).id }} }
+    let(:third_new_coach) { -> {{ name: 'coach', team_id: team.id, user_id: create(:user).id }} }
+    let(:fourth_new_coach) { -> {{ name: 'coach', team_id: team.id, user_id: create(:user).id }} }
+    let(:fifth_new_coach) { -> {{ name: 'coach', team_id: team.id, user_id: create(:user).id }} }
+    let(:remove_coach) { -> {{ name: 'coach', team_id: team.id, user_id: create(:user).id, _destroy: true }} }
+
+    context 'when team has no coaches yet' do
+      it 'allows to add 2 new coaches' do
+        roles_attributes = [new_coach.call, second_new_coach.call, third_new_coach.call, fourth_new_coach.call]
+        expect {
+          team.update roles_attributes: roles_attributes
+        }.to change { team.members.count }.by 4
+      end
+
+      it 'ignores coaches marked for destruction' do
+        roles_attributes = [new_coach.call, second_new_coach.call, third_new_coach.call, fourth_new_coach.call]
+        roles_attributes << remove_coach.call
+        expect {
+          team.update roles_attributes: roles_attributes
+        }.to change { team.members.count }.by 4
+      end
+
+      it 'does not allow to add more than 5 new coaches' do
+        roles_attributes = [new_coach.call, second_new_coach.call, third_new_coach.call, fourth_new_coach.call, fifth_new_coach.call]
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:roles].first).to eql 'there cannot be more than 4 coaches on a team.'
+      end
+
+      it 'does not allow the same coach to fill up both spots' do
+        roles_attributes = [new_coach.call] * 2
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:base].first).to eql "#{new_user.name} can't have more than one role in this team!"
+      end
+
+      it 'does not allow the same coach to add themselves as a student' do
+        roles_attributes = [new_coach.call, new_coach_as_student.call]
+        team.attributes = { roles_attributes: roles_attributes }
+        expect { team.save }.not_to change { team.members.count }
+        expect(team.errors[:base].first).to eql "#{new_user.name} can't have more than one role in this team!"
+      end
+    end
+    context 'when team has coaches' do
+      let!(:coach_1) { create :coach_role, team: team, user: create(:user) }
```

Comment:
You can probably use FactoryBot's  `create_list` for that

Reply:
The only problem I've encountered here is that this would apply the same instance of `User` to all attributes. I'll look into a way to solve this.

Reply:
Instead, using `4.times.map { create :coach_role, team: team, user: create(:user) }` does the trick. What do you think?